bridge/tracker/tracker.go

```diff
@@ -0,0 +1,307 @@
+//	and starts listening for events on the rootchain.
+func (t *Tracker) Start() error {
+	//	subscribe for new headers
+	if err := t.subscribeToRootchain(); err != nil {
+		t.logger.Error("cannot subscribe to rootchain", "err", err)
+		return err
+	}
+
+	//	start processing new header events
+	go t.startEventTracking()
+
+	return nil
+}
+
+//	Stop stops the tracker's listening mechanism.
+func (t *Tracker) Stop() error {
+	//	stop subscription
+	t.ctxSubscription.cancel()
+
+	//	close db
+	t.db.Close()
+
+	//	close rootchain client
+	if err := t.client.close(); err != nil {
+		t.logger.Error("cannot close rootchain client", "err", err)
+
+		return err
+	}
+
+	return nil
+}
+
+//	getEventChannel returns the tracker's event channel.
+func (t *Tracker) GetEventChannel() <-chan []byte {
+	return t.eventCh
+}
+
+//	subscribeToRootchain subscribes the tracker for new
+//	header events on the rootchain.
+func (t *Tracker) subscribeToRootchain() error {
+	//	create cancellable context for tracker
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	t.ctxSubscription = contextSubscription{
+		context: ctx,
+		cancel:  cancelFunc,
+	}
+
+	//	create subscription for new header events
+	var err error
+	if t.sub, err = t.client.subscribeNewHeads(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//	startEventTracking handles the subscription object (provided by the rootchain client).
+func (t *Tracker) startEventTracking() {
+	for {
+		select {
+		//	process new header
+		case header := <-t.sub.newHead():
+			t.trackHeader(header)
+
+		//	handle sub error
+		case err := <-t.sub.err():
+			t.ctxSubscription.cancel()
+			t.logger.Error("subscription cancelled", err)
+
+			return
+
+		//	stop tracker's sub
+		case <-t.ctxSubscription.done():
+			if err := t.sub.unsubscribe(); err != nil {
+				t.logger.Error("cannot unsubscribe", "err", err)
+
+				return
+			}
+
+			t.logger.Debug("subscription stopped")
+
+			return
+		}
+	}
+}
+
+//	trackHeader determines the range of block to query
+//	based on the blockNumber of the header and issues an appropriate
+//	eth_getLogs call. If an event of interest was emitted,
+//	it is sent to eventCh.
+func (t *Tracker) trackHeader(header *ethHeader) {
+	//	determine range of blocks to query
+	fromBlock, toBlock := t.calculateRange(header)
+	if fromBlock == nil || toBlock == nil {
+		//	we are returning here because the calculated
+		//	range was already queried or the chain is not
+		//	at the desired depth.
+		return
+	}
+
+	t.logger.Info("querying events within block range", "from", fromBlock, "to", toBlock)
+
+	//	fetch logs
+	logs := t.queryEvents(fromBlock, toBlock)
+	t.logger.Info("matched events", "num", len(logs))
+
+	//	notify each matched log event
+	t.notify(logs...)
+}
+
+//	calculateRange determines the next range of blocks
+//	to query for events.
+func (t *Tracker) calculateRange(header *ethHeader) (from, to *big.Int) {
+	//	extract block number from header field
+	//	TODO: polygon-edge header
```

Comment:
@dbrajovic what do these TODOs mean?

Reply:
These TODOs are related to the ethereum vs polygon-edge `header` type. (https://github.com/0xPolygon/polygon-edge/pull/419#discussion_r818495762)